### PR TITLE
get all tokens with pagination 147

### PR DESCRIPTION
**Pull request by lzxm160**
work on #147 
using:
1、
```
query {
  xrc20 {
    xrc20Addresses(pagination:{first:3,skip:1}) {
      exist
      count
      addresses
    }
  }
}
```
2、
```
query Xrc20Query($pagination:Pagination!) {
  xrc20 {
    xrc20Addresses(pagination:$pagination) {
      exist
      count
     addresses
    }
  }
}
```
```
{
  "pagination":{
"first":3,
"skip":0
}
}
```

### CODE REVIEW

File: graphql/resolver.go [hunk 1]
```diff
@@ -777,6 +780,37 @@ func (r *queryResolver) getXrc20ByPage(ctx context.Context, actionResponse *Xrc2
...
+		return errors.Wrap(err, "failed to get pagination arguments for get xrc20 addresses")
+	}
+	skip := paginationMap["skip"]
+	first := paginationMap["first"]
+	if skip < 0 {
+		return ErrPaginationInvalidOffset
+	}
+	if first <= 0 || first > MaximumPageSize {
+		return ErrPaginationInvalidSize
+	}
```
Comment (CoderZhi) [inline graphql/resolver.go]: as you can see there, we have this duplicate logic everywhere

File: queryprotocol/actions/protocol.go [hunk 2]
```diff
@@ -479,6 +486,40 @@ func (p *Protocol) GetXrc20ByPage(offset, limit uint64) (cons []*Xrc20Info, err
...
+
+	var ret Address
+	parsedRows, err := s.ParseSQLRows(rows, &ret)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to parse results")
+	}
+	if len(parsedRows) == 0 {
+		err = indexprotocol.ErrNotExist
+		return nil, err
+	}
```
Comment (CoderZhi) [inline queryprotocol/actions/protocol.go]: why do we return error instead of an empty list?

File: graphql/resolver.go (hunk 1 above)
Comment (lzxm160) [reply graphql/resolver.go]: ok,I'll move these to getPaginationArgs

File: graphql/resolver.go [hunk 3]
```diff
@@ -1142,6 +1146,14 @@ func getPaginationArgs(argsMap map[string]*ast.Value) (map[string]int, error) {
...
 	}
+	offset, ok := paginationMap["skip"]
+	if ok && offset < 0 {
+		return paginationMap, ErrPaginationInvalidOffset
+	}
+	size, ok := paginationMap["first"]
+	if ok && (size <= 0 || size > MaximumPageSize) {
+		return paginationMap, ErrPaginationInvalidSize
+	}
 	return paginationMap, nil
```
Comment (CoderZhi) [inline graphql/resolver.go]: return type of paginationMap could be map[string]uint64

File: graphql/resolver.go [hunk 4]
```diff
@@ -777,6 +756,31 @@ func (r *queryResolver) getXrc20ByPage(ctx context.Context, actionResponse *Xrc2
...
+		return errors.Wrap(err, "failed to get pagination arguments for get xrc20 addresses")
+	}
+	skip := paginationMap["skip"]
+	first := paginationMap["first"]
+	output := &XRC20AddressList{Exist: false}
+	actionResponse.Xrc20Addresses = output
+	addresses, err := r.AP.GetXrc20Addresses(uint64(skip), uint64(first))
+	switch {
+	case errors.Cause(err) == indexprotocol.ErrNotExist:
+		return nil
```
Comment (CoderZhi) [inline graphql/resolver.go]: why should we return nil in this case?

File: graphql/resolver.go (hunk 3 above)
Comment (lzxm160) [reply graphql/resolver.go]: ok

**Pull request merged**